webapp/packages/plugin-holidays/src/Christmas/Christmas.ts

```diff
@@ -0,0 +1,289 @@
+      speed = Math.random() * 1 + 0.33,
+      opacity = Math.random() * 0.5 + 0.3;
+
+    flakes.push({
+      speed: speed,
+      shape: Math.floor(Math.random() * 5 + 5),
+      velY: speed,
+      velX: 0,
+      x: x,
+      y: y,
+      size: size,
+      stepSize: Math.random() / 30,
+      step: 0,
+      angle: Math.random() < 0.5 ? 1 : -1,
+      opacity: opacity,
+    });
+  }
+}
+
+function respawnFlake(flake: Flake, width: number) {
+  flake.y = 0;
+  flake.x = Math.floor(Math.random() * width);
+  flake.speed = Math.random() * 0.5 + 0.4;
+  flake.velY = flake.speed;
+  flake.velX = 0;
+  flake.opacity = Math.random() * 0.5 + 0.3;
+}
+
+const MIN_EFFECTIVE_DISTANCE = 300;
+const FRAME_DURATION = 1000 / 60; // 60 FPS
+const FLAKE_ADDING_INTERVAL = 100; //ms
+const SNOWFALL_TIMEOUT = 60000; // 1 minute
+const BASE_FLAKES_COUNT = 250;
+const BASE_CANVAS_AREA = 1920 * 1080;
+
+export class Christmas {
+  private isMouseMoving = false;
+  private flakes: Flake[] = [];
+  private lastFrameTime = 0;
+  private lastFlakeTime = 0;
+  private mX = -MIN_EFFECTIVE_DISTANCE;
+  private mY = -MIN_EFFECTIVE_DISTANCE;
+  private ctx: CanvasRenderingContext2D | null = null;
+  private canvas: HTMLCanvasElement | null = null;
+  private maxFlakesCount = 0;
+  private stopTimeoutId: number | undefined = undefined;
+  private mouseMovingTimeoutId: number | undefined = undefined;
+  private isRunning = false;
+  public isSnowFalling = false;
+
+  constructor() {
+    makeObservable(this, {
+      isSnowFalling: observable,
+    });
+  }
+
+  calculateMaxFlakesCount(width: number, height: number) {
+    const currentArea = width * height;
+    this.maxFlakesCount = Math.round((currentArea / BASE_CANVAS_AREA) * BASE_FLAKES_COUNT);
+  }
+
+  createCanvas() {
+    this.canvas = document.createElement('canvas');
+    this.canvas.width = window.innerWidth;
+    this.canvas.height = window.innerHeight;
+    this.canvas.style.position = 'fixed';
+    this.canvas.style.top = '0';
+    this.canvas.style.left = '0';
+    this.canvas.style.pointerEvents = 'none';
+    this.canvas.style.zIndex = '1';
+    document.body.appendChild(this.canvas);
+    this.ctx = this.canvas.getContext('2d');
+  }
+
+  handleMouseMove = throttle(() => {
+    this.isMouseMoving = true;
+
+    if (this.mouseMovingTimeoutId) {
+      clearTimeout(this.mouseMovingTimeoutId);
+    }
+    this.mouseMovingTimeoutId = window.setTimeout(() => {
+      this.isMouseMoving = false;
+    }, 100);
+  }, 60);
+
+  onResize = debounce(() => {
+    this.calculateMaxFlakesCount(window.innerWidth, window.innerHeight);
+    if (this.canvas) {
+      this.canvas.width = window.innerWidth;
+      this.canvas.height = window.innerHeight;
+    }
+  }, 500);
+
+  onMouseMove = (e: MouseEvent) => {
+    this.handleMouseMove();
+    this.mX = e.clientX;
+    this.mY = e.clientY;
+  };
+
+  onMouseLeave = () => {
+    this.mX = -MIN_EFFECTIVE_DISTANCE;
+    this.mY = -MIN_EFFECTIVE_DISTANCE;
+  };
```

Comment:
please use `this.onMouseMove = this.onMouseMove.bind(this)` in the constructor instead or arrow function in the component